spring-beans/src/main/java/org/springframework/beans/factory/annotation/AutowiredAnnotationBeanPostProcessor.java

```diff
@@ -644,21 +644,20 @@ protected void inject(Object bean, @Nullable String beanName, @Nullable Property
 				}
 				synchronized (this) {
 					if (!this.cached) {
+						Object cachedFieldValue = null;
 						if (value != null || this.required) {
-							this.cachedFieldValue = desc;
+							cachedFieldValue = desc;
 							registerDependentBeans(beanName, autowiredBeanNames);
 							if (autowiredBeanNames.size() == 1) {
 								String autowiredBeanName = autowiredBeanNames.iterator().next();
 								if (beanFactory.containsBean(autowiredBeanName) &&
 										beanFactory.isTypeMatch(autowiredBeanName, field.getType())) {
-									this.cachedFieldValue = new ShortcutDependencyDescriptor(
+									cachedFieldValue = new ShortcutDependencyDescriptor(
 											desc, autowiredBeanName, field.getType());
 								}
 							}
 						}
-						else {
-							this.cachedFieldValue = null;
-						}
+						this.cachedFieldValue = cachedFieldValue;
```

Comment:
hmm, but `cachedFieldValue` _and_ `cached` is `volatile`? A volatile write (`this.cached = true`) ensures that all _subsequent_ reads that see this write, will also observer a fully and initialized `cachedFieldValue`... what I am missing?

Reply:
Nothing changes here as of seeing `cachedFieldValue`, because 1) we are inside of synchronized block and 2) `cachedFieldValue` is anyway assigned. Here I try to avoid the case when we assign `this.cachedFieldValue = desc;` (line 648 in original code) and then reassign it with `new ShortcutDependencyDescriptor()` (line 654 in original code) which is redundant. Instead we can write the values into an intermediate variable and then write it into volatile field only once

Reply:
may be I was not clear enough, sorry. I perfectly understand what you are trying to do - yes, it also makes sense. No question about it. There is a minor problem with your reasoning (or wording) : indeed you are doing the _write_ in a `synchronized` block, but the _read_ happens outside:

```
if (this.cached) {
	value = resolvedCachedArgument(beanName, this.cachedFieldValue);
}
```

anyway, that is not the point I was trying to make. 

My question is rather different, why keep `cachedFieldValue` as `volatile`? It surely looks like `cached` being `volatile` would be enough.  Does that make sense now?

Reply:
Yes,now I clearly understand your point. I'd keep `cachedFieldValue`volatile because it's not a bottleneck, and there's a risk of breaking the code.